README.md
# jquery-idleTimeout

Highly configurable idle (no activity) timer and logout redirect for jQuery.

**Functions across multiple browser windows, tabs and, optionally, iframes (single or nested) as long as the iframes meet the '[Same-Origin Policy](http://en.wikipedia.org/wiki/Same-origin_policy)'.**

Listed on [NPM](https://www.npmjs.com/package/jquery-idletimeout) and [JQuery's Plugin site](http://plugins.jquery.com/idleTimeout/).

Requires [Marcus Westin's store.js](https://github.com/marcuswestin/store.js) which uses localStorage, globalStorage and userData behavior to 'communicate' across multiple browser windows & tabs without cookies or flash.

**[Basic Demo Page](http://jillelaine.github.io/jquery-idleTimeout/) and [Iframes Demo Page](http://jillelaine.github.io/jquery-idleTimeout/iframe-demo.html)**

**[Informative Wiki Pages!](https://github.com/JillElaine/jquery-idleTimeout/wiki)**

#### Communication Across Multiple Browser Windows, Tabs and Iframes

* Functions across multiple instances of a browser window and across multiple browser tabs within the same domain
* Use **jquery-idleTimeout-iframes.min.js** if detection of activity within 'same domain' iframes is desired

#### Required Dependencies

* JQuery core - version 1.7 or newer
* JQuery UI - version 1.9 or newer
* store.js - [https://github.com/marcuswestin/store.js](https://github.com/marcuswestin/store.js) - version 1.3.4 or newer

#### Functionality

* If a window or tab is logged out, all other windows and tabs will log out too
* If **warning dialog** pops up on a window or tab, a **warning dialog** appears on all other windows and tabs too
* If **'Stay Logged In'** button on **warning dialog** is clicked, warning dialogs on all other windows and tabs will be dismissed too
* If **'Log Out Now'** button on **warning dialog** is clicked, all other windows and tabs will log out too
* Public function for your site's **Logout** button so that all open windows and tabs will logout too
* Keep-alive pings server every 10 minutes (default) to prevent server-side session timeout
* All user message text may be modified to your desired language
* Use the optional configuration variable `customCallback` to add function(s) which execute just before user logout

**If the warning dialog box is enabled:**
* After the `idleTimeLimit` amount of user inactivity, the warning dialog box with 2 buttons appear. Default button may be activated with mouse click or press of Enter key.
* Warning dialog includes countdown 'Time remaining' display.
* Browser window and tab title bar(s) display warning message when warning dialog appears. Original browser title restored to all windows and tabs if warning dialog is dismissed.
* Warning dialog will display for the configured `dialogDisplayLimit` amount of time. If user remains idle, plugin will redirect to configured `redirectUrl`.

![Warning Dialog](https://raw.github.com/JillElaine/jquery-idleTimeout/master/warning_dialog.png)

**If the warning dialog box is disabled:**
* After the configured `idleTimeLimit` amount of user inactivity, idleTimeout script will redirect to configured `redirectUrl`.
* No warning dialog box will appear and browser window and tab title bar(s) do not display a warning.

#### How to Use

* Load the required jQuery dependencies on your website: jQuery core and jQuery UI
* Download [store.min.js](https://github.com/marcuswestin/store.js) and the appropriate minified jquery-idleTimeout script: jquery-idleTimeout.min.js or jquery-idleTimeout-iframes.min.js
* Upload these .js files and make them available to your website
* Call the jquery-idleTimeout script in a 'document ready', **`$(document).ready(function ()...`**, function somewhere on your site
* Configure the `redirectUrl` variable within the 'document ready' function to redirect to your site's logout page
* Use the public logout function on your site's 'Logout' button: **`$.fn.idleTimeout().logout()`**

For lots more information, please read the [Wiki](https://github.com/JillElaine/jquery-idleTimeout/wiki)

Detailed information on all the **[Public Configuration Variables](https://github.com/JillElaine/jquery-idleTimeout/wiki/Public-Configuration-Variables)**.

**[Example Usage Page](https://github.com/JillElaine/jquery-idleTimeout/blob/master/example.html)**

**Example Basic Document Ready Function**

```Javascript
  $(document).ready(function () {
    $(document).idleTimeout({
      redirectUrl:  '/logout' // redirect to this url. Set this value to YOUR site's logout page.
    });
  });
```

**Example Site Logout Button**

If user voluntarily logs out of your site with your 'Logout' button (instead of timing out), you can force all 'same domain' windows and tabs to log out too! Attach this small snippet of code, **`$.fn.idleTimeout().logout();`**, to the 'onclick' function of your site's 'Logout' button. See example below.

```
<input value="Logout" onclick="$.fn.idleTimeout().logout();" type="button" title="Logout ALL Windows & Tabs" />
```

#### Iframe Information

If you require activity detection within iframes, use the **jquery-idleTimeout-iframe.min.js** script. 

Please read the [Iframe Wiki Page](https://github.com/JillElaine/jquery-idleTimeout/wiki/Iframes---Information-&-Troubleshooting).

#### Troubleshooting

Please read the [Troubleshooting Wiki Page](https://github.com/JillElaine/jquery-idleTimeout/wiki/General-Troubleshooting)

##### Possible 'mousemove' bug with Chrome browser (on Windows?)

User g4g4r1n reports 'mousemove' event sometimes fires when mouse is not moving on Chrome browser and offers a [possible solution](https://github.com/JillElaine/jquery-idleTimeout/issues/13).

##### Your suggestions and bug reports are appreciated

Use **jquery-idleTimeout-for-testing.js** with Firefox with Firebug add-on or similar for debugging. Your feedback helps to improve this plugin!

changelog
Changelog - JillElaine/jquery-idleTimeout.js
--------------------------------------------
v.1.0.10
+ Fix 'Time Remaining' dialog timeout bug. Closes issue #17

v.1.0.9
+ Update README for NPM markdown

v.1.0.8
+ Add public functions: logout & iframeRecheck
+ Modify console logs, code and variable names to make plugin more understandable
+ Make idleCheckHeartbeat a public configuration variable
+ Created workaround for bug in public function call: iframeRecheck (used defaultConfig instead of userRuntimeConfig)
+ Changed more complex polling functions to use setTimeout instead of setInterval
+ Add ability to disable warning dialog. Closes issue #8.
+ Change configurable timers' units from milliseconds to seconds
+ Merge pull request #9 from svenbit: destroy dialog function set to idleTimeout dialog ID instead of jQuery dialog class selector
+ Merge pull request #10 from svenbit: fix hide dialog "x" close button to select only jquery-idleTimeout dialog
+ Add ability to translate all display text to desired language. Thanks to Criptos. Closes Pull Request #12.
+ Add information to README about possible Chrome 'mousemove' bug (on Windows?). Thanks to g4g4r1n. Closes Issue #13.
+ Make sessionKeepAliveUrl a public variable. Closes issue #15.
+ Stop session keep-alive pings when warning dialog appears; restart keep-alive if warning dialog is dismissed. Closes issue #16.
+ Add ability to detect iframes that are dynamically added to the document, including iframes in dialogs. Closes issue #14.
+ Add ability to detect and include nested iframes, whether dynamic or static.

v1.0.6
+ Modify sessionKeepAliveTimer default to 10 minutes. Set user's current URL as URL to ping. Closes issue #7.
+ Add ability to detect activity within iframes within the same domain. Closes issue #5.
+ Upgrade all code to JSLint compliant

v1.0.5
+ Add 'Time remaining' countdown to warning dialog. Closes issue #6.
+ Change to MIT License
+ Fix bug in 'isDialogOpen' test

v1.0.4
+ Add modification of browser title(s) and tab(s) when warning dialog appears. Original browser title restored if warning dialog dismissed.
+ Remove trailing comma in 'defaults' array

v1.0.3
+ Clean up of code format to standards http://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml

v1.0.2
+ Complete rewrite of script to support Internet Explorer

v1.0.1
+ Add clearTimeout(idleTimer) to storage_event_reactor
+ Reversed order of dialog buttons. 'Stay logged in' has focus.
+ Move private dialog variable to dialog function
+ Update index.html example page to reference Google hosted JQuery scripts and CSS

v1.0
+ First versioned version
+ Add support for disable of server pings. Closes issue #1.
+ Move 'event listener' to all windows & tabs, not just 'inactive' ones. Closes issue #3.
+ Modify custom callback support - does not override redirect url
+ Add support to disable redirect url
+ Create 'destroy & remove warning dialog' as separate function
+ Add testing version - jquery-idleTimeout-for-testing.js


jquery-idleTimeout-with-countdown.js
/**
 * This work is licensed under the MIT License
 *
 * Configurable idle (no activity) timer and logout redirect for jQuery.
 * Works cross-browser with multiple windows and tabs within the same domain.
 * Countdown timer in modal dialog box with seconds remaining until log out.
 *
 * Dependencies: JQuery v1.7+, JQuery UI, store.js from https://github.com/marcuswestin/store.js - v1.3.4+
 * version 1.0.5
 **/

(function($) {

  $.fn.idleTimeout = function(options) {
    //##############################
    //## Configuration Variables
    //##############################
    var defaults = {
      idleTimeLimit: 1200000,       // 'No activity' time limit in milliseconds. 1200000 = 20 Minutes
      dialogDisplayLimit: 180000,   // Time to display the warning dialog before redirect (and optional callback) in milliseconds. 180000 = 3 Minutes
      redirectUrl: '/logout',       // redirect to this url

      // optional custom callback to perform before redirect
      customCallback: false,       // set to false for no customCallback
      // customCallback:    function() {    // define optional custom js function
          // perform custom action before logout
      // },

      // configure which activity events to detect
      // http://www.quirksmode.org/dom/events/
      // https://developer.mozilla.org/en-US/docs/Web/Reference/Events
      activityEvents: 'click keypress scroll wheel mousewheel mousemove', // separate each event with a space

      //dialog box configuration
      dialogTitle: 'Session Expiration Warning',
      dialogText: 'Because you have been inactive, your session is about to expire.',

      // server-side session keep-alive timer & url
      sessionKeepAliveTimer: 60000, // Ping the server at this interval in milliseconds. 60000 = 1 Minute
      // sessionKeepAliveTimer: false, // Set to false to disable pings.
      sessionKeepAliveUrl: '/'  // url to ping
    };

    //##############################
    //## Private Variables
    //##############################
    var opts = $.extend(defaults, options);
    var idleTimer, dialogTimer, remainingTimer, idleTimerLastActivity;
    var checkHeartbeat = 2000; // frequency to check for timeouts - 2000 = 2 seconds.
    var origTitle = document.title; // save original browser title

    //##############################
    //## Private Functions
    //##############################

    var openWarningDialog = function() {
     var dialogContent = "<div id='idletimer_warning_dialog'><p>" + opts.dialogText + "</p><p style='display:inline'>Time remaining: <div style='display:inline' id='countdownDisplay'></div></p></div>";

      var warningDialog = $(dialogContent).dialog({
        buttons: {
          "Stay Logged In": function() {
            destroyWarningDialog();
            stopDialogTimer();
            startIdleTimer();
          },
          "Log Out Now": function() {
            logoutUser();
          }
        },
        closeOnEscape: false,
        modal: true,
        title: opts.dialogTitle
      });

      // hide the dialog's upper right corner "x" close button
      $('.ui-dialog-titlebar-close').css('display', 'none');

      // start the countdown display
      countdownSecondsDisplay();

      document.title = opts.dialogTitle;
    };

    var isDialogOpen = function() {
      var dialogOpen = $('#idletimer_warning_dialog').dialog('isOpen');

      if (dialogOpen === true) {
        return true;
      } else {
        return false;
      }
    };

    // display remaining time on warning dialog
    var countdownSecondsDisplay = function() {
      var dialogDisplaySeconds = opts.dialogDisplayLimit / 1000;
      var mins, secs;

      remainingTimer = setInterval(function() {
          mins = Math.floor(dialogDisplaySeconds / 60); // minutes
          if (mins < 10) { mins = '0' + mins };
          secs = dialogDisplaySeconds - (mins * 60); // seconds
          if (secs < 10) { secs = '0' + secs };
          $('#countdownDisplay').html(mins + ':' + secs);
          dialogDisplaySeconds -= 1;
        }, 1000);
    };

    var destroyWarningDialog = function() {
      $(".ui-dialog-content").dialog('destroy').remove();
      document.title = origTitle;
    };

    var checkIdleTimeout = function() {
      var timeNow = $.now();
      var timeIdleTimeout = (store.get('idleTimerLastActivity') + opts.idleTimeLimit);

      if (timeNow > timeIdleTimeout) {
        if (isDialogOpen() !== true) {
          openWarningDialog();
          startDialogTimer();
        }
      } else if (store.get('idleTimerLoggedOut') === true) { //a 'manual' user logout?
        logoutUser();
      } else {
        if (isDialogOpen() === true) {
          destroyWarningDialog();
          stopDialogTimer();
        }
      }
    };

    var startIdleTimer = function() {
      stopIdleTimer();
      idleTimerLastActivity = $.now();
      store.set('idleTimerLastActivity', idleTimerLastActivity);
      idleTimer = setInterval(checkIdleTimeout, checkHeartbeat);
    };

    var stopIdleTimer = function() {
      clearInterval(idleTimer);
    };

    var checkDialogTimeout = function() {
      var timeNow = $.now();
      var timeDialogTimeout = (store.get('idleTimerLastActivity') + opts.idleTimeLimit + opts.dialogDisplayLimit);

      if ((timeNow > timeDialogTimeout) || (store.get('idleTimerLoggedOut') === true)) {
        logoutUser();
      }
    };

    var startDialogTimer = function() {
      dialogTimer = setInterval(checkDialogTimeout, checkHeartbeat);
    };

    var stopDialogTimer = function() {
      clearInterval(dialogTimer);
      clearInterval(remainingTimer);
    };

    var logoutUser = function() {
      store.set('idleTimerLoggedOut', true);

      if (opts.customCallback) {
        opts.customCallback();
      }

      if (opts.redirectUrl) {
        window.location.href = opts.redirectUrl;
      }
    };

    var activityDetector = function() {

      $('body').on(opts.activityEvents, function() {

        if (isDialogOpen() !== true) {
          startIdleTimer();
        }
      });
    };

    var keepSessionAlive = function() {

      if (opts.sessionKeepAliveTimer) {
        var keepSession = function() {
          if (idleTimerLastActivity === store.get('idleTimerLastActivity')) {
            $.get(opts.sessionKeepAliveUrl);
          }
        };

        setInterval(keepSession, opts.sessionKeepAliveTimer);
      }
    };

    //###############################
    // Build & Return the instance of the item as a plugin
    // This is your construct.
    //###############################
    return this.each(function() {

      if (store.enabled) {
        idleTimerLastActivity = $.now();
        store.set('idleTimerLastActivity', idleTimerLastActivity);
        store.set('idleTimerLoggedOut', false);
      } else {
        alert('Dependent file missing. Please see: https://github.com/marcuswestin/store.js');
      }

      activityDetector();

      keepSessionAlive();

      startIdleTimer();
    });
  }
})(jQuery);


jquery-idleTimeout-for-testing.js
/**
 * This work is licensed under the Creative Commons Attribution-Share Alike 3.0
 * United States License. To view a copy of this license,
 * visit http://creativecommons.org/licenses/by-sa/3.0/us/ or send a letter
 * to Creative Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.
 *
 * Modified by: Jill Elaine
 * Email: jillelaine01@gmail.com
 *
 * Configurable idle (no activity) timer and logout redirect for jQuery.
 * Works across multiple windows, tabs and iframes from the same domain.
 *
 * Dependencies: JQuery v1.7+, JQuery UI, store.js from https://github.com/marcuswestin/store.js - v1.3.4+
 *
 * Commented and console logged for debugging with Firefox & Firebug or similar
 * version 1.0.10
 **/

/*global jQuery: false, document: false, store: false, clearInterval: false, setInterval: false, setTimeout: false, clearTimeout: false, window: false, alert: false, console: false*/
/*jslint indent: 2, sloppy: true, plusplus: true*/

(function ($) {

  $.fn.idleTimeout = function (userRuntimeConfig) {

    console.log('start of jquery-idleTimeout plugin');

    //##############################
    //## Public Configuration Variables
    //##############################
    var defaultConfig = {
      redirectUrl: '/logout',      // redirect to this url on logout. Set to "redirectUrl: false" to disable redirect

      // idle settings
      idleTimeLimit: 30,           // 30 seconds for testing. 'No activity' time limit in seconds. 1200 = 20 Minutes
      idleCheckHeartbeat: 2,       // Frequency to check for idle timeouts in seconds

      // optional custom callback to perform before logout
      customCallback: false,       // set to false for no customCallback
      // customCallback:    function () {    // define optional custom js function
          // perform custom action before logout
      // },

      // configure which activity events to detect
      // http://www.quirksmode.org/dom/events/
      // https://developer.mozilla.org/en-US/docs/Web/Reference/Events
      activityEvents: 'click keypress scroll wheel mousewheel', // separate each event with a space

      // warning dialog box configuration
      enableDialog: true,           // set to false for logout without warning dialog
      dialogDisplayLimit: 20,       // 20 seconds for testing. Time to display the warning dialog before logout (and optional callback) in seconds. 180 = 3 Minutes
      dialogTitle: 'Session Expiration Warning', // also displays on browser title bar
      dialogText: 'Because you have been inactive, your session is about to expire.',
      dialogTimeRemaining: 'Time remaining',
      dialogStayLoggedInButton: 'Stay Logged In',
      dialogLogOutNowButton: 'Log Out Now',

      // error message if https://github.com/marcuswestin/store.js not enabled
      errorAlertMessage: 'Please disable "Private Mode", or upgrade to a modern browser. Or perhaps a dependent file missing. Please see: https://github.com/marcuswestin/store.js',

      // server-side session keep-alive timer
      sessionKeepAliveTimer: 600,   // ping the server at this interval in seconds. 600 = 10 Minutes. Set to false to disable pings
      sessionKeepAliveUrl: window.location.href // set URL to ping - does not apply if sessionKeepAliveTimer: false
    },

    //##############################
    //## Private Variables
    //##############################
      currentConfig = $.extend(defaultConfig, userRuntimeConfig), // merge default and user runtime configuration
      origTitle = document.title, // save original browser title
      storeConfiguration, // public function support, store private configuration variables
      activityDetector,
      startKeepSessionAlive, stopKeepSessionAlive, keepSession, keepAlivePing, // session keep alive
      idleTimer, remainingTimer, checkIdleTimeout, checkIdleTimeoutLoop, startIdleTimer, stopIdleTimer, // idle timer
      openWarningDialog, dialogTimer, checkDialogTimeout, startDialogTimer, stopDialogTimer, isDialogOpen, destroyWarningDialog, countdownDisplay, // warning dialog
      logoutUser,
      checkForIframes, includeIframes, attachEventIframe; // iframes

    //##############################
    //## Public Functions
    //##############################
    // trigger a manual user logout
    // use this code snippet on your site's Logout button: $.fn.idleTimeout().logout();
    this.logout = function () {
      console.log('start logout');
      store.set('idleTimerLoggedOut', true);
    };

    // trigger a recheck for iframes
    // use this code snippet after an iframe is inserted into the document: $.fn.idleTimeout().iframeRecheck()
    this.iframeRecheck = function () {
      console.log('start iframeRecheck');
      checkForIframes();
    };

    //##############################
    //## Private Functions
    //##############################
    //----------- PUBLIC FUNCTION SUPPORT --------------//
    // WORK AROUND - save currentConfig.activityEvents value to 'store.js' variable for use in function: attachEventIframe
    storeConfiguration = function () {
      console.log('store configuration - currentConfig.activityEvents ' + currentConfig.activityEvents + '.');
      store.set('activityEvents', currentConfig.activityEvents);
    };

    //----------- KEEP SESSION ALIVE FUNCTIONS --------------//
    startKeepSessionAlive = function () {
      console.log('start startKeepSessionAlive');

      keepSession = function () {
        console.log('startKeepSessionAlive - send ping to sessionKeepAliveUrl');
        $.get(currentConfig.sessionKeepAliveUrl);
        startKeepSessionAlive();
      };

      keepAlivePing = setTimeout(keepSession, (currentConfig.sessionKeepAliveTimer * 1000));
    };

    stopKeepSessionAlive = function () {
      console.log('stop keep session alive function');
      clearTimeout(keepAlivePing);
    };

    //----------- ACTIVITY DETECTION FUNCTION --------------//
    activityDetector = function () {

      $('body').on(currentConfig.activityEvents, function () {

        if (!currentConfig.enableDialog || (currentConfig.enableDialog && isDialogOpen() !== true)) {
          console.log('activity detected');
          startIdleTimer();
        } else {
          console.log('warning dialog open. activity ignored');
        }
      });
    };

    //----------- IDLE TIMER FUNCTIONS --------------//
    checkIdleTimeout = function () {

      var timeIdleTimeout = (store.get('idleTimerLastActivity') + (currentConfig.idleTimeLimit * 1000));

      if ($.now() > timeIdleTimeout) {
        console.log('inactivity has exceeded the idleTimeLimit');

        if (!currentConfig.enableDialog) { // warning dialog is disabled
          console.log('warning dialog disabled - log out user without warning');
          logoutUser(); // immediately log out user when user is idle for idleTimeLimit
        } else if (currentConfig.enableDialog && isDialogOpen() !== true) {
          console.log('warning dialog is not open & will be opened');
          openWarningDialog();
          startDialogTimer(); // start timing the warning dialog
        }
      } else if (store.get('idleTimerLoggedOut') === true) { //a 'manual' user logout?
        console.log('user has manually logged out? Log out all windows & tabs now.');
        logoutUser();
      } else {
        console.log('inactivity has not yet exceeded the idleTimeLimit');

        if (currentConfig.enableDialog && isDialogOpen() === true) {
          console.log('warning dialog is open & will be closed');
          destroyWarningDialog();
          stopDialogTimer();
        }
      }
    };

    startIdleTimer = function () {
      console.log('start startIdleTimer');
      stopIdleTimer();
      store.set('idleTimerLastActivity', $.now());
      checkIdleTimeoutLoop();
    };

    // continually check if user inactivity has exceeded the idleTimeLimit
    checkIdleTimeoutLoop = function () {
      checkIdleTimeout();
      idleTimer = setTimeout(checkIdleTimeoutLoop, (currentConfig.idleCheckHeartbeat * 1000));
    };

    stopIdleTimer = function () {
      console.log('start stopIdleTimer');
      clearTimeout(idleTimer);
    };

    //----------- WARNING DIALOG FUNCTIONS --------------//
    openWarningDialog = function () {
      console.log('start openWarningDialog');

      var dialogContent = "<div id='idletimer_warning_dialog'><p>" + currentConfig.dialogText + "</p><p style='display:inline'>" + currentConfig.dialogTimeRemaining + ": <div style='display:inline' id='countdownDisplay'></div></p></div>";

      $(dialogContent).dialog({
        buttons: [{
          text: currentConfig.dialogStayLoggedInButton,
          click: function () {
            console.log('Stay Logged In button clicked');
            destroyWarningDialog();
            stopDialogTimer();
            startIdleTimer();
          }
        },
          {
            text: currentConfig.dialogLogOutNowButton,
            click: function () {
              console.log('Log Out Now button clicked');
              logoutUser();
            }
          }
          ],
        closeOnEscape: false,
        modal: true,
        title: currentConfig.dialogTitle,
        open: function () {
          //hide the dialog's upper right corner "x" close button
          $(this).closest('.ui-dialog').find('.ui-dialog-titlebar-close').hide();
        }
      });

      // start the countdown display
      countdownDisplay();

      // change title bar to warning message
      document.title = currentConfig.dialogTitle;

      // if keep-alive is enabled, stop the session keep-alive ping
      if (currentConfig.sessionKeepAliveTimer) {
        stopKeepSessionAlive();
      }
    };

    checkDialogTimeout = function () {
      var timeDialogTimeout = (store.get('idleTimerLastActivity') + (currentConfig.idleTimeLimit * 1000) + (currentConfig.dialogDisplayLimit * 1000));

      if (($.now() > timeDialogTimeout) || (store.get('idleTimerLoggedOut') === true)) {
        console.log('warning dialog is open and user has remained inactive for the dialogDisplayLimit. Time to log out user.');
        logoutUser();
      } else {
        console.log('dialog not yet timed out');
      }
    };

    startDialogTimer = function () {
      console.log('start startDialogTimer');
      dialogTimer = setInterval(checkDialogTimeout, (currentConfig.idleCheckHeartbeat * 1000));
    };

    stopDialogTimer = function () {
      clearInterval(dialogTimer);
      clearInterval(remainingTimer);
    };

    isDialogOpen = function () {
      var dialogOpen = $("#idletimer_warning_dialog").is(":visible");

      if (dialogOpen === true) {
        return true;
      }
      return false;
    };

    destroyWarningDialog = function () {
      console.log('start destroyWarningDialog');
      $("#idletimer_warning_dialog").dialog('destroy').remove();
      document.title = origTitle;

      // if keep-alive is enabled, restart the session keep-alive ping 
      if (currentConfig.sessionKeepAliveTimer) {
        startKeepSessionAlive();
      }
    };

    // display remaining time on warning dialog
    countdownDisplay = function () {
      var dialogDisplaySeconds = currentConfig.dialogDisplayLimit, mins, secs;

      remainingTimer = setInterval(function () {
        mins = Math.floor(dialogDisplaySeconds / 60); // minutes
        if (mins < 10) { mins = '0' + mins; }
        secs = dialogDisplaySeconds - (mins * 60); // seconds
        if (secs < 10) { secs = '0' + secs; }
        $('#countdownDisplay').html(mins + ':' + secs);
        dialogDisplaySeconds -= 1;
      }, 1000);
    };

    //----------- LOGOUT USER FUNCTION --------------//
    logoutUser = function () {
      console.log('start logoutUser');
      store.set('idleTimerLoggedOut', true);

      if (currentConfig.sessionKeepAliveTimer) {
        stopKeepSessionAlive();
      }

      if (currentConfig.customCallback) {
        console.log('logout function custom callback');
        currentConfig.customCallback();
      }

      if (currentConfig.redirectUrl) {
        console.log('logout function redirect to URL');
        window.location.href = currentConfig.redirectUrl;
      }
    };

    //----------- IFRAME FUNCTIONS --------------//
    // triggered when a dialog is opened, recheck for iframes
    $("body").on("dialogopen", function () {
      console.log('start body dialogopen');

      if (currentConfig.enableDialog && isDialogOpen() !== true) {
        console.log('a dialog (not the idleTimeout warning dialog) opened. Recheck for iframes.');
        checkForIframes();
      } else {
        console.log('warning dialog opened. No recheck for iframes.');
      }
    });

    // document must be in readyState 'complete' before checking for iframes - $(document).ready() is not good enough!
    checkForIframes = function () {
      console.log('start checkForIframes');

      var docReadyCheck, isDocReady;

      docReadyCheck = function () {
        if (document.readyState === "complete") {
          console.log('check for iframes, now that the document is complete');
          clearInterval(isDocReady);
          includeIframes();
        }
      };

      isDocReady = setInterval(docReadyCheck, 1000); // check once a second to see if document is complete
    };

    // find and include iframes
    includeIframes = function (elementContents) {
      console.log('start includeIframes');

      if (!elementContents) {
        console.log('elementContents not defined. Define as $(document)');
        elementContents = $(document);
      }

      var iframeCount = 0;

      elementContents.find('iframe,frame').each(function () {
        console.log('start of find iframes');

        if ($(this).hasClass('jit-inspected') === false) {

          console.log('first time inpection of iframe - no jit-inspected class');

          try {

            includeIframes($(this).contents()); // recursive call to include nested iframes

            // attach event code for most modern browsers
            $(this).on('load', attachEventIframe($(this))); // Browser NOT IE < 11

            // attach event code for older Internet Explorer browsers
            console.log('iframeCount: ' + iframeCount + '.');
            var domElement = $(this)[iframeCount]; // convert jquery object to dom element

            if (domElement.attachEvent) { // Older IE Browser < 11
              console.log('attach event to iframe. Browser IE < 11');
              domElement.attachEvent('onload', attachEventIframe($(this)));
            }

            iframeCount++;

          } catch (err) {
            /* Unfortunately, if the 'includeIframes' function is manually triggered multiple times in quick succession,
             * this 'try/catch' block may not have time to complete,
             * and so it might error out, 
             * and iframes that are NOT cross-site may be set to "cross-site"
             * and activity may or may not bubble to parent page.
             * Fortunately, this is a rare occurrence!
             */
            console.log('found cross-site iframe - add classes "jit-inspected" & "cross-site"');
            $(this).addClass('jit-inspected cross-site');
          }

        } else {
          console.log('iframe already has class jit-inspected');
        }

      });

    };

    // attach events to each iframe
    attachEventIframe = function (iframeItem) {
      console.log('start attachEventIframe');

      // retrieve stored value as currentConfig will not include private userRuntimeConfig
      // when this function is called by public function, iframeRecheck
      console.log('currentConfig.activityEvents: ' + currentConfig.activityEvents + '.');
      var iframeContents = iframeItem.contents(), storeActivityEvents = store.get('activityEvents');

      try {

        iframeContents.on(storeActivityEvents, function (event) {
          console.log('bubbling iframe activity event to body of page event: ' + event.type + '.');
          $('body').trigger(event);
        });

        iframeItem.addClass('jit-inspected'); // add "jit-inspected" class, so we don't need to check this iframe again

      } catch (err) {
        console.log('problem with attachment of activity events to this iframe');
      }

    };

    //###############################
    // Build & Return the instance of the item as a plugin
    // This is your construct.
    //###############################
    return this.each(function () {

      if (store.enabled) {

        store.set('idleTimerLastActivity', $.now());
        store.set('idleTimerLoggedOut', false);

        storeConfiguration();

        activityDetector();

        if (currentConfig.sessionKeepAliveTimer) {
          startKeepSessionAlive();
        }

        startIdleTimer();

        checkForIframes();

      } else {
        console.log('store.js not enabled, or browser "private mode"?');
        alert(currentConfig.errorAlertMessage);
      }

    });
  };
}(jQuery));